### DISCUSSION TITLE
Override the `parentbased_traceIdRatio` sampled decision

### DISCUSSION THREAD

**Discussion by itsmaneesh** [open-telemetry/opentelemetry-java-instrumentation#8809] [Q&A]
Currently, we are using `parentbased_traceIdRatio` sampler and '0.3' (30 %) ratio configured as env variables and the sampler is working fine by sampling probabilistic 30% of traces and dropping the remaining 70% traces.

Is there a way to override the sampled decision of a request (trace) on conditional based check.
For e.g.: 
Assume that a request flows from Service1 > Service2 > Service3  (provided all of them are already auto-instrumented and all of them have the same sampling strategy configured.

Is there way to override the sampled decision at any stage (of the above services) during the request execution? If so, can you please provide pointers to override the sampled decision for e.g.: from `drop` to `sample` (record).

If overriding the sampled decision is not possible then can you please provide your recommendations to this use case.

**Comment by mateuszrzeszutek** [comment]
You should be able to do that by implementing a custom sampler that makes the decision you want, or delegates to the `parentbased_traceIdRatio` sampler if the condition did not apply.

A word of warning though: if you decide to start sampling in the middle of an otherwise `drop`ped trace, you will end up sending disjoint & incomplete trace data to your observability backend. It might or might not be understood and rendered properly.

**Reply by itsmaneesh** [reply]
@mateuszrzeszutek Are you saying that I can have a custom sampler to make sample decision based on the condition and if that condition didn't apply then I can fall back to `parentbased_traceIdRatio` sampler that I would have already configured using env variable? If so, can you provide any sample code (reference / Unit test code) that I can refer to add such configuration for my use case.

I understand that if I decide to start (override) the sampling decision in the middle of `drop`ped trace then I will end up seeing incomplete trace data.

**Comment by trask** [comment]
> If so, can you provide any sample code (reference / Unit test code) that I can refer to add such configuration for my use case.

If you are using the Java agent, then you will need to write an extension in order to provide your own custom sampler, check out https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension
